File: tests/render_lcd.cpp
#include "../components/charger_display/charger_display.h"
#include <iostream>
using namespace esphome::charger_event_bus;
using namespace esphome::charger_display;
void emit(const char *name, const Snapshot &state, uint32_t now = 1000) {
  const View view = make_view(state, now);
  std::cout << "PAGE\t" << name << '\n';
  for (size_t i = 0; i < view.count; ++i) {
    const auto &label = view.labels[i];
    std::cout << label.x << '\t' << label.y << '\t' << static_cast<int>(label.font) << '\t'
        << static_cast<int>(label.ink) << '\t' << label.right << '\t' << label.text << '\n';
  }
}
int main() {
  Snapshot state;
  state.connected = state.connection_enabled = state.ready = state.ui_buttons_ready = true;
  state.voltage = 58.4f;
  state.current = 5.1f;
  state.ui_voltage = 58.3f;
  state.ui_current = 5.1f;
  emit("lcd-output-preview.png", state);
  state.ui_mode = UiMode::EDIT;
  emit("lcd-edit-preview.png", state);
  state.ui_mode = UiMode::CONFIRM;
  emit("lcd-confirm-preview.png", state);
  state.ui_mode = UiMode::HELP;
  emit("lcd-help-preview.png", state);
  state.ui_mode = UiMode::VIEW;
  state.telemetry_supported = state.telemetry_valid = state.telemetry_seen = true;
  state.sampled_at = 1000;
  state.output_voltage = 53.8f;
  state.output_current = 4.9f;
  emit("lcd-live-simulation.png", state);
  emit("lcd-stale-simulation.png", state, 8000);
  state.connected = state.connection_enabled = state.ready = false;
  state.telemetry_valid = false;
  emit("lcd-disconnected-preview.png", state);
}

File: tests/render_lcd.py
"""Render documented LCD previews from the firmware's shared C++ view model.

These are layout previews (not photos or measured charging evidence). The
normal view shows a provisional voltage fixture and unavailable current.
Font rasterization uses Pillow; coordinates, text, sizes and colors are shared
with the actual firmware. Requires Pillow, available with ESPHome.
"""
import argparse
import os
from pathlib import Path
import subprocess
import tempfile
from PIL import Image, ImageDraw, ImageFont
import yaml

ROOT = Path(__file__).resolve().parents[1]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--sdk')
    parser.add_argument('--output', type=Path, default=ROOT / 'docs/images')
    args = parser.parse_args()
    with tempfile.TemporaryDirectory(prefix='lxy-lcd-render-') as work:
        binary = str(Path(work) / 'render')
        command = [os.environ.get('CXX', 'c++'), '-std=c++17', '-Wall', '-Wextra', '-Werror']
        if args.sdk:
            command += ['-isysroot', args.sdk]
        subprocess.run(command + [str(ROOT / 'tests/render_lcd.cpp'), '-o', binary], check=True)
        lines = subprocess.check_output([binary], text=True).splitlines()
    declared = set(yaml.safe_load((ROOT / 'assets/ui-glyphs.yaml').read_text()))
    labels = (ROOT / 'components/charger_display/charger_display.h').read_text()
    missing = {c for c in labels if ord(c) > 127} - declared
    if missing:
        raise RuntimeError(f'UI font inventory missing characters: {sorted(missing)}')
    fonts = [ImageFont.truetype(str(ROOT / 'assets/ChargerSansSC.ttf'), n) for n in (12, 14)] + [ImageFont.truetype(str(ROOT / 'assets/Roboto.ttf'), n) for n in (40, 76, 28)]
    colors = [(255, 255, 255), (160, 160, 160), (64, 230, 140), (255, 185, 64)]
    names = []
    args.output.mkdir(parents=True, exist_ok=True)
    pages = []
    for line in lines:
        parts = line.split('\t')
        if parts[0] == 'PAGE':
            names.append(parts[1])
            pages.append(Image.new('RGB', (240, 135), 'black'))
            continue
        x, y, font, ink, right = map(int, parts[:5])
        text = parts[5]
        # ESPHome TOP text uses the font's top line box, not the glyph's ink box.
        draw = ImageDraw.Draw(pages[-1])
        anchor = 'ra' if right else 'la'
        box = draw.textbbox((x, y), text, font=fonts[font], anchor=anchor)
        if box[0] < 0 or box[1] < 0 or box[2] > 240 or box[3] > 135:
            raise RuntimeError(f'Text outside 240x135 screen: {text!r}: {box}')
        draw.text((x, y), text, font=fonts[font], fill=colors[ink], anchor=anchor)
    for name, page in zip(names, pages, strict=True):
        page.resize((960, 540), Image.Resampling.NEAREST).save(args.output / name)
        print(name)


if __name__ == '__main__':
    main()
